### PR TITLE
All feedbacks makeover

### PR DESCRIPTION
**Pull request by maureensturgeon**
On the all feedbacks page, groups feedback by level, displays only the latest feedback for each level by default. The user can see the full history of feedback given on the level by clicking "Show older comments"

Screen Shot 2021-09-07 at 5 54 38 PM

## Links

- jira ticket: [LP-1989](https://codedotorg.atlassian.net/browse/LP-1989)

## Testing story
Tested locally and added unit tests

## PR Checklist:

- [ ] Tests provide adequate coverage
- [ ] Privacy and Security impacts have been assessed
- [ ] Code is well-commented
- [ ] New features are translatable or updates will not break translations
- [ ] Relevant documentation has been added or updated
- [ ] User impact is well-understood and desirable
- [ ] Pull Request is labeled appropriately
- [ ] Follow-up work items (including potential tech debt) are tracked and linked

### CODE REVIEW

File: dashboard/app/controllers/teacher_feedbacks_controller.rb [hunk 1]
```diff
@@ -20,24 +20,18 @@ def index
...
-    marked_latest_for_level = {}
+    @feedbacks_by_level = []
 
-    @feedbacks_as_student_with_level_info = @feedbacks_as_student.map do |feedback|
-      level_feedback_key = [feedback.level_id, feedback.script_id].join('_')
-      is_latest = !marked_latest_for_level[level_feedback_key]
-
-      if is_latest
-        marked_latest_for_level[level_feedback_key] = true
+    @feedbacks_as_student.group_by {|feedback| "#{feedback.level_id}_#{feedback.script_id}"}.each do |_, feedbacks|
```
Comment (jamescodeorg) [inline dashboard/app/controllers/teacher_feedbacks_controller.rb]: (minor) For the group by clause, would using `"#{feedback.script_id}_#{feedback.level_id}"` be equivalent?  If so, I prefer using script first where possible to reflect the hierarchical order.

Comment (jamescodeorg) [inline dashboard/app/controllers/teacher_feedbacks_controller.rb]: This block is a little hard to parse.  Maybe try putting the `.group_by` and `.each` on different lines and/or reduce the block nesting?

File: dashboard/app/views/teacher_feedbacks/index.html.haml [hunk 2]
```diff
@@ -1,5 +1,5 @@
 - feedback_data = {}
-- feedback_data[:all_feedback] = @feedbacks_as_student_with_level_info
+- feedback_data[:all_feedback_by_level] = @feedbacks_by_level
```
Comment (jamescodeorg) [inline dashboard/app/views/teacher_feedbacks/index.html.haml]: Do we want to use `feedback` or `feedbacks` as the plural form?  (I don't have a strong opinion of which we use as long as we are consistent.)

File: apps/src/templates/feedback/LevelFeedback.jsx [hunk 3]
```diff
@@ -0,0 +1,98 @@
+import React, {Component} from 'react';
+import i18n from '@cdo/locale';
+import color from '@cdo/apps/util/color';
+import {levelFeedbackType} from './types';
+import LevelFeedbackEntry from '@cdo/apps/templates/feedback/LevelFeedbackEntry';
+import Button from '@cdo/apps/templates/Button';
+
+export default class LevelFeedback extends Component {
```
Comment (jamescodeorg) [inline apps/src/templates/feedback/LevelFeedback.jsx]: This is totally fine, but just a reminder that functional components with hooks are also an option now!

File: apps/src/templates/feedback/LevelFeedback.jsx [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    } = this.props;
+
+    const {showingOlderComments} = this.state;
+
+    const displayedFeedbacks = showingOlderComments
+      ? feedbacks
+      : feedbacks.slice(0, 1);
+
+    const hasMultipleFeedbacks = feedbacks.length > 1;
+    const marginBottom = hasMultipleFeedbacks ? 8 : 24;
```
Comment (jamescodeorg) [inline apps/src/templates/feedback/LevelFeedback.jsx]: Why do we need this?

**Review by jamescodeorg — COMMENTED** [review]
I think this will be huge improvement for students!

File: dashboard/app/controllers/teacher_feedbacks_controller.rb (hunk 1 above)
Comment (maureensturgeon) [reply dashboard/app/controllers/teacher_feedbacks_controller.rb]: Yes, should be the same, I'll update.

File: apps/src/templates/feedback/LevelFeedback.jsx (hunk 4 above)
Comment (maureensturgeon) [reply apps/src/templates/feedback/LevelFeedback.jsx]: For levels that have just one feedback we do not display the "Show older comments" button, so the design has a little bit of extra spacing to make up for the space that we usually get with the button.

File: apps/src/templates/feedback/LevelFeedback.jsx (hunk 3 above)
Comment (maureensturgeon) [reply apps/src/templates/feedback/LevelFeedback.jsx]: HOOKS! I forgot, this seems like a good opportunity to start using them, I'll convert.

File: apps/src/templates/feedback/LevelFeedbackEntry.jsx [hunk 5]
```diff
@@ -13,196 +14,170 @@ const getElementHeight = element => {
...
+    review_state,
+    seen_on_feedback_page_at,
+    student_first_visited_at,
+    is_awaiting_teacher_review
+  } = feedback;
+
+  const feedbackSeenByStudent =
+    seen_on_feedback_page_at || student_first_visited_at;
+
+  const commentExists = comment.length > 2;
```
Comment (jamescodeorg) [inline apps/src/templates/feedback/LevelFeedbackEntry.jsx:35]: Do you know why this is `>2`?  (I see that it is logic carried over from the previous implementation.)

File: dashboard/db/schema.rb [hunk 6]
```diff
@@ -10,7 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2021_08_20_182004) do
+ActiveRecord::Schema.define(version: 2021_09_08_195236) do
```
Comment (jamescodeorg) [inline dashboard/db/schema.rb]: The changes to schema.rb and schema_cache.yml are not related to this change, right?

File: dashboard/test/controllers/teacher_feedbacks_controller_test.rb [hunk 7]
```diff
@@ -69,6 +100,6 @@ class TeacherFeedbacksControllerTest < ActionController::TestCase
   def get_all_response_feedback_data
     assert_select 'script[data-feedback]', 1
     feedback_data = JSON.parse(css_select('script[data-feedback]').first.attribute('data-feedback').to_s)
-    feedback_data['all_feedback']
+    feedback_data['all_feedback_by_level']
```
Comment (jamescodeorg) [inline dashboard/test/controllers/teacher_feedbacks_controller_test.rb]: Should this be `all_feedbacks_by_level`?

File: dashboard/test/controllers/teacher_feedbacks_controller_test.rb [hunk 8]
```diff
@@ -14,8 +14,8 @@ class TeacherFeedbacksControllerTest < ActionController::TestCase
     get :index
     assert_response :success
 
-    all_feedback_data = get_all_response_feedback_data
-    assert_equal 0, all_feedback_data.count
+    all_feedback_by_level = get_all_response_feedback_data
```
Comment (jamescodeorg) [inline dashboard/test/controllers/teacher_feedbacks_controller_test.rb]: I think this should probably be `all_feedbacks_by_level` to be consistent with the rest of the naming?

File: dashboard/app/controllers/teacher_feedbacks_controller.rb [hunk 9]
```diff
@@ -20,24 +20,21 @@ def index
...
-      is_latest = !marked_latest_for_level[level_feedback_key]
+    feedbacks_grouped_by_level = @feedbacks_as_student.group_by {|feedback| "#{feedback.script_id}_#{feedback.level_id}"}
+    
+    feedbacks_grouped_by_level.each do |_, feedbacks|
+      level_details = feedbacks[0].get_script_level&.summary_for_feedback
 
-      if is_latest
-        marked_latest_for_level[level_feedback_key] = true
+      summarized_feedbacks = feedbacks.each_with_index.map do |feedback, i|
+        isLatest = i == 0
```
Comment (jamescodeorg) [inline dashboard/app/controllers/teacher_feedbacks_controller.rb:31]: Nice!

**Review by jamescodeorg — APPROVED** [review]
Love it!  I'm finding the UI display logic to be a lot more localized and easier to understand.

File: apps/src/templates/feedback/LevelFeedbackEntry.jsx (hunk 5 above)
Comment (maureensturgeon) [reply apps/src/templates/feedback/LevelFeedbackEntry.jsx:35]: I'm not sure. Must have been an issue at some point that there were accidental or empty string comments?

File: dashboard/db/schema.rb (hunk 6 above)
Comment (maureensturgeon) [reply dashboard/db/schema.rb]: Oops! Yes, I'll remove that, good catch. That was related to my other migration PR

**Review by maureensturgeon — COMMENTED** [review]
Feel free to merge while I'm out.
